feat(docs): update Aztec.nr contract structure and state variables (#19726)
AztecProtocol/aztec-packages: docs/docs-developers/docs/aztec-nr/framework-description/state_variables.md

#### Declaration

Store mutable public state using `PublicMutable<T>` for values that need to be updated throughout the contract's lifecycle.

For example, storing the address of the collateral asset in a lending contract:

#include_code public_mutable /noir-projects/noir-contracts/contracts/app/lending_contract/src/main.nr rust

#### `read`

`PublicMutable` variables have a `read` method to read the value at the location in storage:

#include_code public_mutable_read /noir-projects/noir-contracts/contracts/app/lending_contract/src/main.nr rust

#### `write`

The `write` method on `PublicMutable` variables takes the value to write as an input and saves this in storage:

#include_code public_mutable_write /noir-projects/noir-contracts/contracts/app/lending_contract/src/main.nr rust

### PublicImmutable

> `PublicImmutable` is a simplified version `PublicMutable`: it's a public state variable that can only be written (initialized) once, at which point it can only be read. Unlike Solidity `immutable` state variables, which must be set in the contract's constructor, a `PublicImmutable` can be initialized _at any point in time_ during the contract's lifecycle and attempts to read it prior to initialization will revert.

Reviewer: ```suggestion
`PublicImmutable` is a simplified version of `PublicMutable`: it's a public state variable that can only be written (initialized) once, at which point it can only be read. Unlike Solidity `immutable` state variables, which must be set in the contract's constructor, a `PublicImmutable` can be initialized _at any point in time_ during the contract's lifecycle and attempts to read it prior to initialization will revert.
```

- Reading the current value results in the state variable being updated, increasing tx costs and requiring delivery of a note message.
- There is no `write` function - the current value is instead `replace`d.

:::info
To ensure that a user's private execution always uses the latest value of a `PrivateMutable`, the `get_note` function will nullify the note that it is reading. This means that if two people are trying to use this function with the same note, only one will succeed.

Reading a `PrivateMutable` nullifies and recreates the note. This makes reads indistinguishable from writes and ensures the sequencer cannot learn the note's value.
:::

#### Declaration

#include_code owned_private_mutable /noir-projects/noir-contracts/contracts/app/app_subscription_contract/src/main.nr rust

#### `is_initialized`

An unconstrained method to check whether the `PrivateMutable` has been initialized or not:

#include_code owned_private_mutable_is_initialized /noir-projects/noir-contracts/contracts/app/app_subscription_contract/src/main.nr rust

#### `initialize` and `initialize_or_replace`

> The `PrivateMutable` should be initialized to create the first note and value. This can be does eith either `initialize` or `initialize_or_replace`:

Reviewer: ```suggestion
The `PrivateMutable` should be initialized to create the first note and value. This can be done with either `initialize` or `initialize_or_replace`:
```

#include_code owned_private_mutable_initialize /noir-projects/noir-contracts/contracts/app/app_subscription_contract/src/main.nr rust

#### `get_note`

This function allows us to get the note of a `PrivateMutable`, essentially reading the value. Note that because :

```rust
#[external("private")]
fn read_settings() {
    let owner = self.msg_sender().unwrap();

>     self.storage.user_settings.at(owner).get_note().deliver(MessageDelivery.CONSTRAINED_ONCHAIN);

Reviewer: ```suggestion
    self.storage.user_settings.at(owner).get_note().deliver(MessageDelivery.ONCHAIN_CONSTRAINED);
```

}
```

:::info

> To ensure that a user's private execution always uses the latest value of a `PrivateMutable`, the `get_note` function will nullify the note that it is reading. This means that if two people are trying to use this function with the same note, only one will succeed.

Reviewer: This block is redundant with lines 285-289. Think it should be removed?

:::

#### `replace`

To update the value of a `PrivateMutable`, we can use the `replace` method:

#include_code owned_single_private_mutable_replace /noir-projects/noir-contracts/contracts/app/private_token_contract/src/main.nr rust

### PrivateImmutable

`PrivateImmutable` represents a unique private state variable that, as the name suggests, is immutable. Once initialized, its value cannot be altered. This is the private equivalent of `PublicImmutable`, except the value is only known to its owner.

Unlike a `PrivateMutable`, the `get_note` function for a `PrivateImmutable` doesn't nullify the current note and returns the `Note` directly (not wrapped in `NoteMessage`). This means that multiple accounts can concurrently call this function to read the value.

#### Declaration

#include_code private_immutable /noir-projects/noir-contracts/contracts/test/test_contract/src/main.nr rust

`PrivateImmutable` variables also have the `initialize` and `get_note` functions on them but no `initialize_or_replace` since they cannot be modified.

> Unlike a `PrivateMutable`, the `get_note` function for a `PrivateImmutable` doesn't nullify the current note and returns the `Note` directly (not wrapped in `NoteMessage`). This means that multiple accounts can concurrently call this function to read the value.

Reviewer: Duplicate of line 333

### PrivateSet

`PrivateSet` is used for managing a collection of notes. Like `PrivateMutable`, this is a private state variable that can be modified. There are two key differences:

- A `PrivateSet` is not a single value but a _set_ (a collection) of values (represented by notes)
- Any account can insert values into someone else's set.

The set's current value is the collection of notes in the set that have not yet been nullified. These notes can have any type: they could be nft IDs, representing a user's nft collection, or they might be token amounts, in which case _the sum_ of all values in the set would be the user's current balance.

#### Declaration

For example, to add private token balances to storage:

> #include_code private_set_insert /noir-projects/noir-contracts/contracts/test/pending_note_hashes_contract/src/main.nr rust

Reviewer: this should probably be #include_code private_set to show the storage struct declaration, not the insert operation

### SinglePrivateMutable and SinglePrivateImmutable

For contract-wide private values (not per-owner), use `SinglePrivateMutable` or `SinglePrivateImmutable`. These store exactly one value for the entire contract - a global singleton - rather than separate values per owner.

| Type | Use Case | Access Pattern |
|------|----------|----------------|
| `Owned<PrivateMutable<...>>` | Per-owner private state (like balances) | `.at(owner).get_note()` |
| `SinglePrivateMutable` | Contract-wide singleton (like admin) | `.get_note()` directly |

Since there's only one value at the storage slot, there's no need to specify an owner to look it up:

```rust
#[storage]
struct Storage<Context> {
    admin: SinglePrivateMutable<AddressNote, Context>,
    config: SinglePrivateImmutable<ConfigNote, Context>,
}

// Access directly without .at(owner)
let note_message = self.storage.admin.get_note();
let config = self.storage.config.get_note();
```

When initializing, you still pass an owner address - but this specifies who can decrypt the note, not the storage location:

```rust
// owner_address determines who can see the note, not where it's stored

> self.storage.admin.initialize(note, owner_address).deliver(MessageDelivery.ONCHAIN_CONSTRAINED);

Reviewer: all of these reference should be MessageDelivery.ONCHAIN_UNCONSTRAINED instead of MessageDelivery.UNCONSTRAINED_ONCHAIN

Recipients learning about notes created for them is known as 'note discovery', which is a process Aztec.nr handles efficiently and automatically. However, it does mean that when a note is created, a _message_ with the content of the note is created and needs to be delivered to a recipient via one of multiple means detailed below.

When working with private state variables, many operations return a `NoteMessage<Note>` type rather than the note directly. This is a type-safe wrapper that ensures you explicitly decide how to deliver the note to its recipient.

#### Delivery Methods

Private notes need to be communicated to their recipients so they know the note exists and can use it. The [`NoteMessage`](pathname:///aztec-nr-api/#api_ref_version/noir_aztec/note/struct.notemessage) wrapper forces you to make an explicit choice about how this happens:

>   - `MessageDelivery.ONCHAIN_CONSTRAINED`](pathname:///aztec-nr-api/#api_ref_version/noir_aztec/messages/message_delivery/struct.messagedeliveryenum#structfield.ONCHAIN_UNCONSTRAINED): Verified in the circuit (most secure, but highest cost) - Use when the sender cannot be trusted to deliver correctly (e.g., protocol fees, multisig config updates). **Warning:** Currently [not fully constrained](https://github.com/AztecProtocol/aztec-packages/issues/14565) - the log's tag is unconstrained.

Reviewer: ```suggestion
  - [`MessageDelivery.ONCHAIN_CONSTRAINED`](pathname:///aztec-nr-api/#api_ref_version/noir_aztec/messages/message_delivery/struct.messagedeliveryenum#structfield.ONCHAIN_UNCONSTRAINED): Verified in the circuit (most secure, but highest cost) - Use when the sender cannot be trusted to deliver correctly (e.g., protocol fees, multisig config updates). **Warning:** Currently [not fully constrained](https://github.com/AztecProtocol/aztec-packages/issues/14565) - the log's tag is unconstrained.
```

:::note
Advanced users can change this default behavior by either defining their [own custom note](./how_to_implement_custom_notes.md) hash and nullifier functions, implementing their own state variables, or even accessing the note hash and nullifiers tree directly.
:::

The snippet below shows a contract with two private state variables: an admin address (stored in an `AddressNote`) and a counter of how many calls the admin has made (stored in a `UintNote`). These values will be private and therefore not known except by the accounts that own these notes (the admin). In the `perform_admin_action` private function, the contract checks that it is being called by the correct admin and updates the call count by incrementing it by one.

(Note that this is not a real snippet, it's missing some small irrelevant details - but the gist of it is correct)

```rust
#[note]
struct AddressNote {
    value: AztecAddress,
}

#[note]
struct UintNote {
    value: u128,
}

#[storage]
struct Storage {
    admin: PrivateImmutable<AddressNote>,

>     admin_call_count: PrivateMutable<UintNote>,

Reviewer: wrap this privatemutable in `Owned`

}

#[external("private")]
fn perform_admin_action() {
    // Read the contract's admin address and check against the caller

>     let admin = self.storage.admin.read().value;

Reviewer: ```suggestion
    let admin = self.storage.admin.get_note().value;
```
because its private immutable

    assert(self.msg_sender() == admin);
    // Update the call count by replacing (updating - rename soon) the current note with a new one that equals the
    // current value + 1 - this requires knowing what the current value is in the first place, i.e., reading the variable.
    //
    // We then deliver the encrypted message with the note's content to the admin so that they become aware of the new
    // value of the counter and can update it again in the future.
    self.storage.admin_call_count
        .replace(|current| UintNote{ value: current.value + 1 }) // wouldn't it be great if we didn't have to deal with this wrapping and unwrapping?
        .deliver(admin);

>     ...

Reviewer: ```suggestion
    // ...
```
just to make it valid syntax

## Choosing the right storage type

| Need | Use |
|------|-----|
| Public value anyone can read/write | `PublicMutable` |
| Public value set once (contract name, decimals) | `PublicImmutable` |
| Public key-value mapping | `Map<K, PublicMutable<V>>` |
| Private collection per user (token balances) | `Owned<PrivateSet<...>>` |
| Single private value per user | `Owned<PrivateMutable<...>>` |
| Immutable private value per user | `Owned<PrivateImmutable<...>>` |
| Contract-wide private singleton (admin key) | `SinglePrivateMutable` |
| Public value readable in private execution | `DelayedPublicMutable` |

## Prerequisites

- An Aztec contract project set up with `aztec-nr` dependency
- Understanding of Aztec's private and public state model
- Familiarity with Noir struct syntax

For storage concepts, see [storage overview](../../foundational-topics/state_management.md).

## The Storage Struct

State variables are declared in Solidity by simply listing them inside of the contract, like so:

```solidity
contract MyContract {
    uint128 public my_public_state_variable;
}
```

> In Aztec.nr, we define a [`struct`](https://noir-lang.org/docs/noir/concepts/data_types/structs) that holds _all_ state variables. This struct is called **the storage struct**, and it is identified by having the `#[storage]` macro applied to it.

Reviewer 2: link to storage macro apiref somewhere

> In Aztec.nr, we define a [`struct`](https://noir-lang.org/docs/noir/concepts/data_types/structs) that holds _all_ state variables. This struct is called **the storage struct**, and it is identified by having the [`#[storage]` macro](https://docs.aztec.network/aztec-nr-api/nightly/noir_aztec/macros/storage/fn.storage) applied to it.

Reviewer: this should be a relative link to the appropriate version. wdyt about adding a preprocessing macro, something like (to `include_version.js`?
                                                                                            
```js
  // New macro for API ref paths                                                                             
  const apiRefVersion = releaseType === "nightly" ? "nightly"                                                
                      : releaseType === "devnet" ? "devnet"                                                  
                      : "next";                                                                              
  markdownContent = markdownContent.replaceAll(`#api_ref_version`, apiRefVersion);                           
```            
                                                                                  
  Then links could be:                                                                                       
 "[`#[storage]` macro] (/aztec-nr-api/#api_ref_version/noir_aztec/macros/storage/fn.storage)"

Author: This won't work bc the nightly & devnet stuff isn't preprocessed is it?? That's what claude says when it broke haha

Below is a table comparing the key properties of the different public state variables that Aztec.nr offers:

| State variable         | Mutable?            | Readable in private? | Writable in private? | Example use case                                                                   |
| ---------------------- | ------------------- | -------------------- | -------------------- | ---------------------------------------------------------------------------------- |
| `PublicMutable`        | yes                 | no                   | no                   | Configuration of admins, global state (e.g. token total supply, total votes)       |
| `PublicImmutable`      | no                  | yes                  | no                   | Fixed configuration, one-way actions (e.g. initialization settings for a proposal) |
| `DelayedPublicMutable` | yes (after a delay) | yes                  | no                   | Non time sensitive system configuration                                            |

> ### PublicMutable

Reviewer 2: apiref linkssss here and elsewhere, for each of these

### Note Messages and Discovery

Because notes are private, not even the intended recipient is aware of their existence, and therefore they must be somehow notified. For example, when making a payment and creating a note for the payee with the intended amount, they must be shown the preimage of the note that was inserted in the note hash tree in a given transaction in order to acknowledge the payment.

Recipients learning about notes created for them is known as 'note discovery', which is a process Aztec.nr handles efficiently and automatically. However, it does mean that when a note is created, a _message_ with the content of the note is created and needs to be delivered to a recipient via one of multiple means detailed below.

When working with private state variables, many operations return a `NoteMessage<Note>` type rather than the note directly. This is a type-safe wrapper that ensures you explicitly decide how to deliver the note to its recipient.

> #### Delivery Methods

Reviewer 2: linky link

Private notes need to be communicated to their recipients so they know the note exists and can use it. The [`NoteMessage`](https://docs.aztec.network/aztec-nr-api/nightly/noir_aztec/note/struct.notemessage) wrapper forces you to make an explicit choice about how this happens:

> [

Reviewer: ```suggestion
```
